fri/src/two_adic_pcs.rs

```diff
@@ -0,0 +1,539 @@
+use alloc::vec;
+use alloc::vec::Vec;
+use core::marker::PhantomData;
+
+use itertools::{izip, Itertools};
+use p3_challenger::{CanSample, FieldChallenger};
+use p3_commit::{DirectMmcs, OpenedValues, Pcs, UnivariatePcs, UnivariatePcsWithLde};
+use p3_dft::TwoAdicSubgroupDft;
+use p3_field::{
+    batch_multiplicative_inverse, cyclic_subgroup_coset_known_order, AbstractField, ExtensionField,
+    Field, PackedField, TwoAdicField,
+};
+use p3_interpolation::interpolate_coset;
+use p3_matrix::bitrev::{BitReversableMatrix, BitReversedMatrixView};
+use p3_matrix::dense::RowMajorMatrixView;
+use p3_matrix::{Dimensions, Matrix, MatrixRows};
+use p3_maybe_rayon::prelude::*;
+use p3_util::linear_map::LinearMap;
+use p3_util::{log2_strict_usize, reverse_bits_len, reverse_slice_index_bits, VecExt};
+use serde::{Deserialize, Serialize};
+use tracing::{info_span, instrument};
+
+use crate::verifier::{self, VerificationErrorForFriConfig};
+use crate::{prover, FriConfig, FriProof};
+
+pub struct TwoAdicFriPcs<FC, Val, Dft, M> {
+    fri: FC,
+    dft: Dft,
+    mmcs: M,
+    _phantom: PhantomData<Val>,
+}
+
+impl<FC, Val, Dft, M> TwoAdicFriPcs<FC, Val, Dft, M> {
+    pub fn new(fri: FC, dft: Dft, mmcs: M) -> Self {
+        Self {
+            fri,
+            dft,
+            mmcs,
+            _phantom: PhantomData,
+        }
+    }
+    fn coset_shift(&self) -> Val
+    where
+        Val: TwoAdicField,
+    {
+        Val::generator()
+    }
+}
+
+#[derive(Debug)]
+pub enum VerificationError<FC: FriConfig, InputMmcsError> {
+    FriError(VerificationErrorForFriConfig<FC>),
+    InputMmcsError(InputMmcsError),
+}
+
+#[derive(Serialize, Deserialize)]
+pub struct TwoAdicFriPcsProof<FC: FriConfig, Val, InputMmcsProof> {
+    #[serde(bound = "")]
+    pub(crate) fri_proof: FriProof<FC>,
+    /// For each query, for each committed batch, query openings for that batch
+    pub(crate) query_openings: Vec<Vec<BatchOpening<Val, InputMmcsProof>>>,
```

Comment:
Hm why split this off, is it that `FriProof` will be shared with circle FRI, but not `query_openings`?

Reply:
Yeah, the input openings are all handled by the PCS wrapper.

Same kind of deal as above: 

> I mean, this is basically the whole reason for the old virtual MMCS, but I think it's worth reifying the complexity here for efficiency gains. Plus, FRI is more of a backend thing and will be immediately wrapped by one or two PCS's that take care of it.

Old way was definitely more elegant and the "right" way to do it, but we're breaking abstraction here a bit, assuming it will only dealt with by the PCS wrappers.